storagemarket/provider_test.go

```diff
@@ -0,0 +1,563 @@
+}
+
+type HarnessOpt func(h *ProviderHarness)
+
+// configures the min publish balance for each deal
+func MinPublishFee(minPublishBal abi.TokenAmount) HarnessOpt {
+	return func(h *ProviderHarness) {
+		h.MinPublishFees = minPublishBal
+	}
+}
+
+// configures the max bytes allocated to the staging area
+func MaxStagingDealBytes(maxBytes uint64) HarnessOpt {
+	return func(h *ProviderHarness) {
+		h.MaxStagingDealBytes = maxBytes
+	}
+}
+
+func NewHarness(t *testing.T, ctx context.Context, opts ...HarnessOpt) *ProviderHarness {
+	// tempDir
+	dir := t.TempDir()
+
+	// setup mocks
+	ctrl := gomock.NewController(t)
+	fn := lotusmocks.NewMockFullNode(ctrl)
+	bp := smtestutil.NewMinerStub(ctrl)
+
+	// setup client and miner addrs
+	minerAddr, err := address.NewIDAddress(1011)
+	require.NoError(t, err)
+	cAddr, err := address.NewIDAddress(1014)
+	require.NoError(t, err)
+
+	// instantiate the http servers that will serve the files
+	normalServer, err := testutil.HttpTestUnstartedFileServer(t, dir)
+	require.NoError(t, err)
+
+	// create a provider libp2p peer
+	mn := mocknet.New(ctx)
+	h, err := mn.GenPeer()
+	require.NoError(t, err)
+
+	// setup the databases
+	sqldb, err := db.CreateTmpDB(ctx)
+	require.NoError(t, err)
+	dealsDB := db.NewDealsDB(sqldb)
+
+	// publish wallet
+	pw, err := address.NewIDAddress(uint64(rand.Intn(100)))
+	require.NoError(t, err)
+
+	// create the harness with default values
+	ph := &ProviderHarness{
+		GoMockCtrl:          ctrl,
+		TempDir:             dir,
+		MinerAddr:           minerAddr,
+		ClientAddr:          cAddr,
+		NormalServer:        normalServer,
+		MockFullNode:        fn,
+		DealsDB:             dealsDB,
+		FundsDB:             db.NewFundsDB(sqldb),
+		StorageDB:           db.NewStorageDB(sqldb),
+		PublishWallet:       pw,
+		MinerStub:           bp,
+		MinPublishFees:      abi.NewTokenAmount(100),
+		MaxStagingDealBytes: 10000000000,
+	}
+
+	// apply the config options
+	for _, o := range opts {
+		o(ph)
+	}
+
+	// fund manager
+	fminitF := fundmanager.New(fundmanager.Config{
+		PubMsgBalMin: ph.MinPublishFees,
+		PubMsgWallet: pw,
+	})
+	fm := fminitF(fn, sqldb)
+
+	// storage manager
+	fsRepo, err := repo.NewFS(dir)
+	require.NoError(t, err)
+	lr, err := fsRepo.Lock(repo.StorageMiner)
+	require.NoError(t, err)
+	smInitF := storagemanager.New(storagemanager.Config{
+		MaxStagingDealsBytes: ph.MaxStagingDealBytes,
+	})
+	sm := smInitF(lr, sqldb)
+	prov, err := NewProvider("", h, sqldb, dealsDB, fm, sm, fn, bp, address.Undef, bp, nil, bp)
+	require.NoError(t, err)
+	prov.testMode = true
+	ph.Provider = prov
+
+	return ph
+}
+
+func (h *ProviderHarness) Start(t *testing.T, ctx context.Context) {
+	h.NormalServer.Start()
+	require.NoError(t, h.Provider.Start(ctx))
+}
+
+func (h *ProviderHarness) Stop() {
+	h.GoMockCtrl.Finish()
+	h.NormalServer.Close()
+}
+
+func (h *ProviderHarness) mkDealProposal(t *testing.T, serverURL string, seed int) (dp *types.DealParams, carV2FilePath string) {
```

Comment:
Nit: I suggest `newDealProposal` as a better name

```suggestion
func (h *ProviderHarness) newDealProposal(t *testing.T, serverURL string, seed int) (dp *types.DealParams, carV2FilePath string) {
```

Reply:
done.